### ISSUE TITLE
Dynamic dependency resolves to a snapshot instead of the latest release

### ISSUE THREAD

**Issue by frost13it** [gradle/gradle#8126]
*(Previously raised as [GRADLE-3097](https://issues.gradle.org/browse/GRADLE-3097))*

Dynamic version range (e.g. `[1,2)` or `1.+`) includes SNAPSHOT versions.

### Expected Behavior

Given that requested version is `[1,2)` and available candidate versions are `1.2` and `1.3-SNAPSHOT` Gradle selects last release version `1.2`.

### Current Behavior

Given that requested version is `[1,2)` and available candidate versions are `1.2` and `1.3-SNAPSHOT` Gradle selects last version `1.3-SNAPSHOT`.

### Context

I need to use the last release of an artifact (with specified major version).
An API or workaround to exclude SNAPSHOT versions from range would be sufficient for me.

**Comment by GFriedrich** [comment]
There is already a proper solution to achieve this:
Please check out the "componentSelection" of the "resolutionStrategy" (see also https://docs.gradle.org/current/dsl/org.gradle.api.artifacts.ComponentSelectionRules.html).
This is by the way also mentioned in a comment of the ticket you've mentioned with an example: https://issues.gradle.org/browse/GRADLE-3097?focusedCommentId=21217&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-21217

**Comment by marcphilipp** [comment]
Closing with the above rationale.

**Comment by frost13it** [comment]
@GFriedrich I've seen it. But that way to find the dependency declaration for candidate in `componnentSelection` via `configuration.dependencies.find` doesn't seem to be reliable.
It looks like it does not work for extended configurations.
I use `configuration.allDependencies.find` and it works in my case. But I am afraid that it can work wrong in some situtations (transitive dependencies etc).
Is there a better way to do this?

**Comment by abishek3876** [comment]
I think this should be supported natively instead of relying on workarounds which use internal APIs. Being able to resolve released versions (or ignoring snapshot or beta builds) in dynamic versions is a common requirement.
@marcphilipp is there any chance this issue could be reopened as a feature request?

**Comment by xenoterracide** [comment]
would it be possible to get a DSL e.g. just being able to call something like this

```
configurations {
    create("rejectConfig") {
        resolutionStrategy {
            componentSelection {
               rejectSnapshotsForDynamicVersions() // because otherwise I have to figure out the magic arcane incantation that is not simply if ( module version ) repeated
            }
        }
    }
}
```

I would also like to note, given the existing behavior which is unlikely to change given how long it has been around... it's hard to tell whether the resolved SNAPSHOT is treated as a changing version, or a dynamic version cache wise.

**Comment by AishD3890** [comment]
Hi is this resolved or still only manual workaround has to be implemented.

**Comment by marcphilipp** [comment]
@gradle/jvm Could one of you guys please take a look?

**Comment by ellisdc** [comment]
Thought I'd post what we did here in case it helps others. A little background:

We generally use some sort of wildcard version for our in-house dependencies. This introduces challenges when dealing with snapshots, since out of the box, gradle will often resolve to snapshot versions due to the wildcard even though we don't always want that. Our rule, is that if you want to use a snapshot dependency, you have to declare an explicit dependency on that snapshot. Therefore, we want to reject snapshot dependencies when resolving against a wildcard.

We initially tried to do this via the aforementioned method:

`configurations.all {
        resolutionStrategy {
            componentSelection.all{
                if (candidate.version.endsWith('-SNAPSHOT')) {
                    def dependency = configurations.allDependencies.find { it.group == candidate.group && it.name == candidate.module }				
                    if (dependency && !dependency.version.endsWith('-SNAPSHOT')) {
                        reject "only releases are allowed for ${dependency}"
                    }
                }
            }
        }
    }`

As @frost13it  mentioned above, this can fail for certain extended configurations AND it doesn't work for transitive dependencies, which makes it a no-go solution for most everyone.

The solution we came up with was to use a combination of methods provided by the [ResolutionStrategy api](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/ResolutionStrategy.html). The componentSelection.all() method only got us half of the way there. It will tell you anytime you have a snapshot candidate, but what it doesn't give you is a link back to the unresolved dependency declaration so that we can determine if a wildcard is in play (to the gradle devs, this would be a great addition to that particular api). 

The eachDependency method does the other half of what we needed. It iterates through all the dependencies, including transitive dependencies, pre resolution. What it lacks is any sort of pointer to the resolution candidates. What we were able to do, though is use eachDependency() to get a list of all dependency declarations pre resolutions, store those in a list, and then use componentSelection.all() to determine when a snapshot is being considered and then search the list that we collected from eachDependency() to see if any of those declarations (which, again, include transitives) are explicitly declared as snapshot dependencies. Otherwise, we reject the snapshot candidate.

`configurations.all {
        List<ModuleVersionSelector> declaredDependencies = []
        resolutionStrategy.eachDependency { DependencyResolveDetails details ->
            declaredDependencies.add(details.getRequested())
        }
    
        resolutionStrategy {
            componentSelection.all {
                if (candidate.version.endsWith('-SNAPSHOT')) {
                    def dependencies = declaredDependencies.findAll { it.group == candidate.group && it.name == candidate.module }
                        
                    if (dependencies && dependencies.find{ it.version.endsWith('-SNAPSHOT') } == null) {
                        reject "only releases are allowed for ${dependency}"
                    }
                }
            }
        }
    }`

**Comment by OndraZizka** [comment]
This should be a default behavior. This is an example of why Gradle is such a hostile build system. Seriously - do I need to write code to get a common sense behavior?

**Comment by tnielens** [comment]
Commenting in support of the proposal above. Most build tools across programming languages have evolved in the direction of version ranges in the main build file and a lockfile for reproducibility. "release" versions should take precedence over "integration" ones by default. For example `1.3.2` should take precedence over `1.3.3-SNAPSHOT` and `1.3.3-alpha1` by default. This way of managing dependencies is very helpful.

**Comment by xenoterracide** [comment]
> 1.3.2 should take precedence over 1.3.3-SNAPSHOT

Did you mean 1.3.3? for varying definitions of precedence. Not if you say that you want a pre-release

General sorting of versions here would be (latest last)
1.3.2
1.3.3-alpha1
1.3.3-SNAPSHOT

at least according to semver

However, I'd expect that unless I explicitly request pre-releases that I don't get them.

big problem of course is that maven doesn't follow semver... although in your example those are valid semvers

1.3.3.FINAL is a prerelease per semver, but not in maven, and would be considered lower in the sort (less stable?)

I think what we kind of need is a way to easily say no pre-release and a way to say what algorithm is used for that (maven vs semver vs ...)

**Comment by tnielens** [comment]
Indeed `1.3.2` comes before `1.3.3-alpha1` as per [version ordering](https://docs.gradle.org/current/userguide/single_versions.html#version_ordering). What this issue and others are suggesting - by default exclude non-release versions, from dynamic version selection - makes a lot of sense to me.

**Comment by xenoterracide** [comment]
I agree with that, was just clarifying... I think. Problem is needing to define pre-release rules and exclude by default.

**Comment by melix** [comment]
I think you are mixing 2 independent dimensions. Dynamic versions are one thing, "quality" of releases is another. I think trying to solve the problem of only getting releases by changing how ranges are interpreted is doomed to fail, because there is simply no relationship between the 2. Instead, there are assumptions which have to be made explicit. For example, "a version which contains -SNAPSHOT" is integration, "a version which contains "BETA" is pre-release", etc. Then, you have rules to say that you do not allow a requested dependency to be downgraded to something of lower release state. For example, I think it's acceptable to be upgraded from 1.1-alpha to 1.2, but it's not to be upgraded to 1.1 to 1.2-alpha. There are all the tools you need in Gradle to do this, using component metadata rules and component selection rules. I'm pretty sure there are tests around this in the codebase, but it's been several years since I last looked so my memory isn't super fresh.

**Comment by tnielens** [comment]
I agree that redefining a dynamic version precedence order next to the existing version order is not the right approach. Exclusion rules make more sense.

And there seems to be ways to achieve this as posted here above. Two suggestions I'd make:
- Consider excluding integration/pre-release versions (`snapshot`, `dev`, `alpha`, etc) by default when the component selector is a dynamic version as this might be the most common use case. Or document this use case. Happy to contribute the documentation if there is a simple solution for this. See next point.
- Make the `ComponentSelector` available in the `ComponentSelection`. Currently in [component selection](https://docs.gradle.org/current/userguide/dynamic_versions.html#sec:component_selection_rules), it's not possible to access the component selector for which a component selection takes place. Therefore it's not possible to exclude pre-releases only for dynamic versions as one might want to configure a pre-release strictly in the dependencies. There are workarounds like walking the `configurations` dependencies and matching on the module identifier, but that seems odd and potentially ineffective.

**Comment by OndraZizka** [comment]
@melix , isn't there already a special behavior of enumerated strings? The logic of some strings meaning pre-release etc is already there, just done half-way: Gradle sorts 2.0-dev before 2.0, but 2.0-foo after 2.0. Still, if you ask for (1, 2), Gradle will happily serve 1.1-SNAPSHOT, 1.1-dev, rather than 1.0. Yes, it is a different matter, but consistent it is not. Ymmv, but I would expect those sorted before the un-suffixed to be optionally disregarded for resolving of ranges.

**Comment by boolivar** [comment]
I would like to add that this stops being a problem if snapshots are located in separate repository.
In this case there is an option to include snapshot repository for versions lookup or use releases only.

**Comment by big-guy** [comment]
We think this is possible with existing APIs without requiring a component selection rule. e.g., it should be almost as simple as setting a property on the Configuration you want to resolve. 

We haven't confirmed it, but we think adding the org.gradle.status=release as an attribute to the resolvable configuration will cause Gradle to ignore SNAPSHOT dependencies in dynamic version ranges. 

We need to investigate and then document this as the solution.
